fairlearn/grid_search.py

```diff
@@ -0,0 +1,127 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""
+This module implements grid searches for fairness/accuracy tradeoffs. Each
+point in the grid corresponds to a different accuracy/fairness trade-off. The
+dimensionality of the search increases with the number of possible values
+for the protected attribute, so for now only binary protected attributes are
+considered
+"""
+
+import copy
+import numpy as np
+
+class BinaryProtectedAttributeDemographicParity:
+    def __init__(self):
+        # Nothing to do
+        return
+
+    def _generate_p0_p1(self, protected_attribute):
+        """ Function to compute p0 and p1 for the given
+        set of protected_attribute values
+        """
+        unique_labels, counts = np.unique(protected_attribute, return_counts=True)
+        if len(unique_labels) > 2:
+            raise RuntimeError("Protected Attribute is not binary")
+
+        p0 = counts[0] / len(protected_attribute)
+        # It is possible that our data only contain one value for
+        # the protected attribute
+        # We do not consider this an error
+        if len(counts) > 1:
+            p1 = counts[1] / len(protected_attribute)
+        else:
+            p1 = 0
+
+        return p0, p1
+
+    def _weight_function(self, y_val, a_val, L, p_ratio):
+        if a_val == 0:
+            return 2*y_val - 1 - L * p_ratio
+        else:
+            return 2*y_val - 1 + L
+
+    def _generate_weights(self, y, protected_attribute, L, p_ratio):
+        weight_func = np.vectorize(self._weight_function)
+        return weight_func(y, protected_attribute, L, p_ratio)
+
+    def grid_search_binary_classification(self, learner, x, y, protected_attribute, lagrange_multipliers=None, number_lagrange_multipliers=11):
```

Comment:
Looking at the OneNote again, I'm wondering if `demographic_parity` should be down here

Reply:
Done